File: test/Microsoft.ServiceHub.Framework.Tests/TestAssets/MockServiceBroker.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.IO.Pipelines;
using Microsoft.ServiceHub.Framework;
using Nerdbank.Streams;

public class MockServiceBroker : IServiceBroker
{
	public event EventHandler<BrokeredServicesChangedEventArgs>? AvailabilityChanged;

	public ValueTask<IDuplexPipe?> GetPipeAsync(ServiceMoniker serviceMoniker, ServiceActivationOptions options, CancellationToken cancellationToken)
	{
		if (TestServices.Calculator.Moniker.Equals(serviceMoniker))
		{
			(IDuplexPipe, IDuplexPipe) pair = FullDuplexStream.CreatePipePair();
			TestServices.Calculator.ConstructRpc(new Calculator(), pair.Item1);
			return new ValueTask<IDuplexPipe?>(pair.Item2);
		}

		if (TestServices.Throws.Moniker.Equals(serviceMoniker))
		{
			throw new ServiceActivationFailedException(serviceMoniker, new Exception("Service factory throws."));
		}

		return default;
	}

	public ValueTask<T?> GetProxyAsync<T>(ServiceRpcDescriptor serviceDescriptor, ServiceActivationOptions options, CancellationToken cancellationToken)
		where T : class
	{
		throw new NotImplementedException();
	}

	internal void OnAvailabilityChanged(BrokeredServicesChangedEventArgs args) => this.AvailabilityChanged?.Invoke(this, args);
}

File: src/servicebroker-npm/test/testAssets/mockServiceBroker.ts
import CancellationToken from 'cancellationtoken'
import { EventEmitter } from 'events'
import { ServiceRpcDescriptor } from '../../src/ServiceRpcDescriptor'
import { IServiceBroker, ServiceBrokerEmitter } from '../../src/IServiceBroker'
import { FullDuplexStream } from 'nerdbank-streams'
import { Calculator } from './calculatorService'
import { calcDescriptorUtf8Http } from './testUtilities'
import { ServiceActivationOptions } from '../../src/ServiceActivationOptions'
import { IDisposable } from '../../src/IDisposable'
import { ServiceMoniker } from '../../src/ServiceMoniker'

export class MockServiceBroker extends (EventEmitter as new () => ServiceBrokerEmitter) implements IServiceBroker {
	getProxy<T extends object>(
		serviceDescriptor: ServiceRpcDescriptor,
		options?: ServiceActivationOptions,
		cancellationToken?: CancellationToken
	): Promise<(T & IDisposable) | null> {
		throw new Error('Method not implemented.')
	}
	getPipe(serviceMoniker: ServiceMoniker, options?: ServiceActivationOptions, cancellationToken?: CancellationToken): Promise<NodeJS.ReadWriteStream | null> {
		if (ServiceMoniker.equals(calcDescriptorUtf8Http.moniker, serviceMoniker)) {
			const pair = FullDuplexStream.CreatePair()
			calcDescriptorUtf8Http.constructRpc(new Calculator(), pair.first)
			return Promise.resolve(pair.second)
		}

		if (serviceMoniker.name === 'throws') {
			throw new Error('Throws on demand')
		}

		return Promise.resolve(null)
	}
}
